RelationalRecord/src/Relations/Tenant.hs

```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE TemplateHaskell, MultiParamTypeClasses, FlexibleInstances #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module  Relations.Tenant where
+
+import  Types.Tenant as Tenant
+import  Types.DB
+import  Relations.DB
+
+import  Database.Relational.Query
+import  Database.HDBC.Query.TH      (makeRecordPersistableDefault)
+
+import  Data.Int                    (Int32)
+
+
+-- SELECTS
+
+allTenants :: Relation () Tenants
+allTenants = Tenant.tenants
+
+getTenant :: Relation PKey Tenants
+getTenant = relation' . placeholder $ \tenId -> do
+    a       <- query tenants
+    wheres  $ a ! Tenant.id' .=. tenId
+    return  a
+
+
+-- INSERTS
+
+data TenantInsert = TenantInsert
+    { iName         :: String
+    , iFirstName    :: String
+    , iLastName     :: String
+    , iPhone        :: String
+    , iEmail        :: String
+    , iOwnerId      :: Maybe PKey
+    , iBOD          :: String
+    }
+$(makeRecordPersistableDefault ''TenantInsert)
+
+piTenant :: Pi Tenants TenantInsert
+piTenant = TenantInsert
+    |$| Tenant.name'
+    |*| Tenant.firstName'
+    |*| Tenant.lastName'
+    |*| Tenant.phone'
+    |*| Tenant.email'
+    |*| Tenant.ownerId'
+    |*| Tenant.backofficeDomain'
+
+insertTenant :: Insert TenantInsert
+insertTenant = derivedInsert piTenant
+
+
+-- UPDATES
+
+data TenantUpdate = TenantUpdate
+    { uName         :: Maybe String
+    , uFirstName    :: Maybe String
+    , uLastName     :: Maybe String
+    , uPhone        :: Maybe String
+    , uEmail        :: Maybe String
+    , uBOD          :: Maybe String
+    , uStatus       :: Maybe Int32
+    , uOwnerId      :: Maybe (Maybe Int32)
```

Comment:
```Maybe a``` is the HRR derivation of a ```nullable``` type ```a``` on the db.

The conflation problem you adressed can actually not occur: The outer ```Maybe``` signals the presence vs. absence of a value for the update. So, no value assignment in the query takes place whatsoever. The inner ```Maybe``` can then be set to ```Just Nothing``` to update with sql ```NULL```. Thus, the call-sites that never update some field (```Nothing```) stay correct, the ones updating a column that has been made nullable fail to type-check, as it should be.

You're right however that the mechanism is not quite clear at first glance. I've changed that by replacing the outer ```Maybe``` with a different, more expressive type.